packages/tupaia-web/src/features/Matrix.tsx

```diff
@@ -0,0 +1,114 @@
+/**
+ * Tupaia
+ * Copyright (c) 2017 - 2023 Beyond Essential Systems Pty Ltd
+ */
+
+import React from 'react';
+import {
+  MatrixColumnType,
+  MatrixRowType,
+  getIsUsingDots,
+  Matrix as MatrixComponent,
+  Alert,
+} from '@tupaia/ui-components';
+import { MatrixDataColumn, MatrixDataRow, MatrixViewContent } from '../types';
+import { ConditionalPresentationOptions } from '@tupaia/types';
+import styled from 'styled-components';
+
+const NoDataMessage = styled(Alert).attrs({
+  severity: 'info',
+})`
+  width: 100%;
+  margin: 1rem auto;
+  max-width: 24rem;
+`;
+
+// This is a recursive function that parses the rows of the matrix into a format that the Matrix component can use.
+const parseRows = (
```

Comment:
There's a lot going on here but separating it out into parseRows and parseColumns and all the comments help clarify it a lot.